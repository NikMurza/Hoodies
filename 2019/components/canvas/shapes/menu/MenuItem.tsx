import * as React from 'react';
import { Group, Line, Rect, Text } from 'react-konva';
import colors from '@common/styles/constants/colors';
import fonts from '@common/styles/constants/fonts';
import UrlImage from '@common/components/lib/canvas/shapes/url-image/UrlImage';

const menuFont = {
  text: 15,
};

export interface ICanvasMenuItem {
  text?: string;
  icon?: string;
  items?: ICanvasMenuItem[];
  object?: any;
}

interface IProps {
  onClick?: (level: number, index: number, item: ICanvasMenuItem) => void;
  index?: number;
  offsetX?: number;
  offsetY?: number;
  level?: number;
  item?: ICanvasMenuItem;
  activeItems?: number[];
}

class MenuItem extends React.PureComponent<IProps> {
  public static defaultProps: IProps = {
    offsetX: 0,
    offsetY: 0,
    item: {
      items: [],
    },
  };

  private onClick = () => {
    const { item, index, level, onClick } = this.props;
    onClick(level, index, item);
  };

  private onHover = (event) => {
    const shape = event.target;
    if (shape.getClassName() === 'Rect') {
      const children = shape.getParent().getChildren();
      children[1].setFill(colors.hover);
      document.body.style.cursor = 'pointer';
      event.target.getLayer().draw();
    }
  };

  private onMouseOut = (event) => {
    const shape = event.target;
    if (shape.getClassName() === 'Rect') {
      const children = shape.getParent().getChildren();
      children[1].setFill(colors.white);
      document.body.style.cursor = 'default';
      event.target.getLayer().draw();
    }
  };

  public render() {
    const { item, index, level, activeItems, onClick, offsetY, offsetX } = this.props;
    const holderWidth = 235;
    const holderHeight = 40;
    const itemWidth = 235;
    const itemHeight = 40;
    return (
      <Group key={`lvl${level}_${index}`}>
        <Group
          x={level * holderWidth + offsetX}
          y={holderHeight * index + offsetY}
          onClick={this.onClick}
          onmousemove={this.onHover}
          onmouseleave={this.onMouseOut}
        >
          <Rect
            x={0}
            y={0}
            width={holderWidth}
            height={holderHeight}
            fill={colors.white}
            cornerRadius={2}
          />

          <Rect
            x={0}
            y={0}
            width={itemWidth}
            height={itemHeight}
            fill={colors.white}
            stroke={colors.transparent}
            cornerRadius={2}
            shadowColor={colors.shadowColor}
            shadowOffsetY={3}
            shadowBlur={4}
          />

          <UrlImage src={item.icon} x={14} y={10} />

          <Text
            text={item.text}
            x={50}
            y={15}
            fill={colors.text}
            fontSize={menuFont.text}
            fontFamily={fonts.family}
          />
          {item.items && (
            <Line points={[215, 16, 221, 22, 215, 28]} stroke={colors.lightBlue} strokeWidth={2} />
          )}
        </Group>
        {activeItems[level] === index &&
          item.items.map((subItem, subIndex) => (
            <MenuItem
              index={subIndex}
              offsetX={1}
              offsetY={index * holderHeight}
              level={level + 1}
              key={subIndex}
              item={subItem}
              activeItems={activeItems}
              onClick={onClick}
            />
          ))}
      </Group>
    );
  }
}

export default MenuItem;
