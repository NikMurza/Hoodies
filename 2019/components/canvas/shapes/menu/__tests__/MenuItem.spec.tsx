import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import colors from '@common/styles/constants/colors';
import MenuItem from '../MenuItem';

describe('MenuItem component', () => {
  it('render default', async () => {
    expect(
      toJson(
        shallow(
          <MenuItem
            index={1}
            offsetY={1}
            level={1}
            item={{
              text: 'menu1',
              items: [
                {
                  text: 'Menu2',
                },
              ],
            }}
            activeItems={[]}
          />,
        ),
      ),
    ).toMatchSnapshot();
  });

  it('render select', async () => {
    expect(
      toJson(
        shallow(
          <MenuItem
            index={1}
            offsetY={1}
            level={0}
            item={{
              text: 'menu1',
              items: [
                {
                  text: 'Menu2',
                },
              ],
            }}
            activeItems={[1]}
          />,
        ),
      ),
    ).toMatchSnapshot();
  });

  it('_onClick', async () => {
    const onClick = jest.fn();
    const wrapper = shallow(
      <MenuItem
        index={1}
        offsetY={1}
        level={1}
        item={{
          text: 'menu1',
          items: [
            {
              text: 'Menu2',
            },
          ],
        }}
        activeItems={[]}
        onClick={onClick}
      />,
    );
    wrapper.find('Group').get(1).props.onClick();
    expect(onClick).toHaveBeenCalled();
  });

  it('_onHover not rect', async () => {
    const getParent = jest.fn();
    const wrapper = shallow(
      <MenuItem
        index={1}
        offsetY={1}
        level={1}
        item={{
          text: 'menu1',
          items: [
            {
              text: 'Menu2',
            },
          ],
        }}
        activeItems={[]}
      />,
    );
    wrapper
      .find('Group')
      .get(1)
      .props.onmousemove({
        target: {
          getClassName: () => 'not Rect',
          getParent,
        },
      });
    expect(getParent).not.toHaveBeenCalled();
  });

  it('_onHover rect without Line', async () => {
    const setStroke = jest.fn();
    const setFill = jest.fn();
    const draw = jest.fn();
    const wrapper = shallow(
      <MenuItem
        index={1}
        offsetY={1}
        level={1}
        item={{
          text: 'menu1',
          items: [
            {
              text: 'Menu2',
            },
          ],
        }}
        activeItems={[]}
      />,
    );
    wrapper
      .find('Group')
      .get(1)
      .props.onmousemove({
        target: {
          getClassName: () => 'Rect',
          getParent: () => ({
            getChildren: () => [
              {},
              {
                setFill,
              },
            ],
          }),
          getLayer: () => ({
            draw,
          }),
        },
      });

    expect(setFill).toHaveBeenCalledWith(colors.hover);
    expect(draw).toHaveBeenCalled();
    expect(document.body.style.cursor).toEqual('pointer');
  });

  it('_onHover rect with Line', async () => {
    const setStroke1 = jest.fn();
    const setFill = jest.fn();
    const draw = jest.fn();
    const setStroke3 = jest.fn();
    const wrapper = shallow(
      <MenuItem
        index={1}
        offsetY={1}
        level={1}
        item={{
          text: 'menu1',
          items: [
            {
              text: 'Menu2',
            },
          ],
        }}
        activeItems={[]}
      />,
    );
    wrapper
      .find('Group')
      .get(1)
      .props.onmousemove({
        target: {
          getClassName: () => 'Rect',
          getParent: () => ({
            getChildren: () => [
              {},
              {
                setFill,
              },
            ],
          }),
          getLayer: () => ({
            draw,
          }),
        },
      });

    expect(setFill).toHaveBeenCalledWith(colors.hover);
    expect(draw).toHaveBeenCalled();
    expect(document.body.style.cursor).toEqual('pointer');
  });

  it('_onMouseOut not rect', async () => {
    const getParent = jest.fn();
    const wrapper = shallow(
      <MenuItem
        index={1}
        offsetY={1}
        level={1}
        item={{
          text: 'menu1',
          items: [
            {
              text: 'Menu2',
            },
          ],
        }}
        activeItems={[]}
      />,
    );
    wrapper
      .find('Group')
      .get(1)
      .props.onmouseleave({
        target: {
          getClassName: () => 'not Rect',
          getParent,
        },
      });
    expect(getParent).not.toHaveBeenCalled();
  });

  it('_onMouseOut rect without Line', async () => {
    const setStroke = jest.fn();
    const setFill = jest.fn();
    const draw = jest.fn();
    const wrapper = shallow(
      <MenuItem
        index={1}
        offsetY={1}
        level={1}
        item={{
          text: 'menu1',
          items: [
            {
              text: 'Menu2',
            },
          ],
        }}
        activeItems={[]}
      />,
    );
    wrapper
      .find('Group')
      .get(1)
      .props.onmouseleave({
        target: {
          getClassName: () => 'Rect',
          getParent: () => ({
            getChildren: () => [
              {},
              {
                setFill,
              },
            ],
          }),
          getLayer: () => ({
            draw,
          }),
        },
      });

    expect(setFill).toHaveBeenCalledWith(colors.white);
    expect(draw).toHaveBeenCalled();
    expect(document.body.style.cursor).toEqual('default');
  });
});
