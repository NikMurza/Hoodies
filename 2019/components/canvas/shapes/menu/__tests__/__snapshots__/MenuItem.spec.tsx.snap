// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItem component render default 1`] = `
<Group
  key="lvl1_1"
>
  <Group
    onClick={[Function]}
    onmouseleave={[Function]}
    onmousemove={[Function]}
    x={235}
    y={41}
  >
    <Rect
      cornerRadius={2}
      fill="#FFF"
      height={40}
      width={235}
      x={0}
      y={0}
    />
    <Rect
      cornerRadius={2}
      fill="#FFF"
      height={40}
      shadowBlur={4}
      shadowColor="#CCC"
      shadowOffsetY={3}
      stroke="transparent"
      width={235}
      x={0}
      y={0}
    />
    <URLImage
      x={14}
      y={10}
    />
    <Text
      fill="#222B3C"
      fontFamily="OpenSans, sans-serif"
      fontSize={15}
      text="menu1"
      x={50}
      y={15}
    />
    <Line
      points={
        Array [
          215,
          16,
          221,
          22,
          215,
          28,
        ]
      }
      stroke="#23B1FE"
      strokeWidth={2}
    />
  </Group>
</Group>
`;

exports[`MenuItem component render select 1`] = `
<Group
  key="lvl0_1"
>
  <Group
    onClick={[Function]}
    onmouseleave={[Function]}
    onmousemove={[Function]}
    x={0}
    y={41}
  >
    <Rect
      cornerRadius={2}
      fill="#FFF"
      height={40}
      width={235}
      x={0}
      y={0}
    />
    <Rect
      cornerRadius={2}
      fill="#FFF"
      height={40}
      shadowBlur={4}
      shadowColor="#CCC"
      shadowOffsetY={3}
      stroke="transparent"
      width={235}
      x={0}
      y={0}
    />
    <URLImage
      x={14}
      y={10}
    />
    <Text
      fill="#222B3C"
      fontFamily="OpenSans, sans-serif"
      fontSize={15}
      text="menu1"
      x={50}
      y={15}
    />
    <Line
      points={
        Array [
          215,
          16,
          221,
          22,
          215,
          28,
        ]
      }
      stroke="#23B1FE"
      strokeWidth={2}
    />
  </Group>
  <MenuItem
    activeItems={
      Array [
        1,
      ]
    }
    index={0}
    item={
      Object {
        "text": "Menu2",
      }
    }
    key="0"
    level={1}
    offsetX={1}
    offsetY={40}
  />
</Group>
`;
